File: haskell/test/Conformance.hs
{-# LANGUAGE OverloadedStrings #-}

-- |
-- The 12-check binding conformance suite (@docs\/conformance.md@).
--
-- Proves the Haskell binding marshals every value shape across the FFI. It is
-- __not__ a sanitizer test suite — the behavioural cases live in the engine's
-- own tests and run once, in Aether.
--
-- == Why a plain runner and not hspec\/tasty
--
-- hspec and tasty arrive from Hackage, so running them needs a @cabal update@
-- and a network round trip (or a pre-warmed store) before a single assertion
-- executes. The rest of this monorepo's bindings test with whatever is already
-- on the box — @php tests\/conformance.php@, a .NET console runner — so this one
-- does too. The dependency set is @base@ + @bytestring@, both of which ship
-- with GHC, which means @runghc@ can drive the whole suite offline:
--
-- @
--     runghc -isrc -itest test\/Conformance.hs
-- @
--
-- The trade is real and small: no test discovery, no @--match@, no shrinking.
-- For thirty-odd marshalling assertions that costs nothing, and it keeps the
-- binding testable on an air-gapped machine. Swapping in hspec later is
-- mechanical — each @check@ below is one @it@.
--
-- The process exit code is the result: 0 all-pass, 1 any failure.
module Main (main) where

import Control.Exception (SomeException, try)
import Control.Monad (forM_, unless, when)
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as C
import Data.IORef
  ( IORef
  , modifyIORef'
  , newIORef
  , readIORef
  , writeIORef
  )
import Data.List (sort)
import System.Exit (exitFailure, exitSuccess)
import System.IO (hSetEncoding, stdout, utf8)

import HtmlSanitizer
import qualified HtmlSanitizer.Native as N

-- ---------------------------------------------------------------------------
-- A minimal assertion runner
-- ---------------------------------------------------------------------------

-- Failures are accumulated rather than fatal, so one broken marshalling path
-- does not hide the other eleven.
type Failures = IORef [String]

-- | An assertion failure, raised inside a check body and caught by 'check'.
--
-- 'ioError' . 'userError' rather than 'error': it throws an ordinary
-- 'IOException' from 'IO' rather than an imprecise exception from pure code,
-- so it is raised exactly where it is written (no laziness surprises about
-- when a failure actually surfaces) and 'show' renders it without a call
-- stack. Both are in the Prelude, so it costs no import.
assertFail :: String -> IO a
assertFail = ioError . userError

-- | Run one check with a freshly-created sanitizer, so each starts from the
-- engine's defaults, and close it afterwards even if the body throws.
check :: Failures -> String -> (Sanitizer -> IO ()) -> IO ()
check fs name body = checkIO fs name (withSanitizer body)

-- | Run one check that manages its own sanitizer(s).
checkIO :: Failures -> String -> IO () -> IO ()
checkIO fs name body = do
  r <- try body :: IO (Either SomeException ())
  case r of
    Right () -> putStrLn ("  PASS " ++ name)
    Left e -> do
      modifyIORef' fs (++ [name ++ ": " ++ show e])
      putStrLn ("  FAIL " ++ name)
      putStrLn ("       " ++ show e)

eqStr :: String -> B.ByteString -> B.ByteString -> IO ()
eqStr what got want =
  unless (got == want) $
    assertFail $
      what
        ++ ":\n         got  \""
        ++ render got
        ++ "\"\n         want \""
        ++ render want
        ++ "\""

-- | Render a ByteString for a failure message.
--
-- 'C.unpack' maps each BYTE to a Char 0..255, so a UTF-8 payload printed
-- through the UTF-8 handle we set in 'main' would come out as mojibake — the
-- bytes get re-encoded a second time. For a diagnostic that is worse than
-- useless when check 4 is the one failing, so any non-ASCII byte switches the
-- whole string to an explicit hex dump.
render :: B.ByteString -> String
render bs
  | B.all (< 0x80) bs = C.unpack bs
  | otherwise = unwords (map hex2 (B.unpack bs))
  where
    digits = "0123456789abcdef"
    hex2 w =
      let n = fromIntegral w :: Int
       in [digits !! (n `div` 16), digits !! (n `mod` 16)]

eqInt :: String -> Int -> Int -> IO ()
eqInt what got want =
  unless (got == want) $
    assertFail (what ++ ": got " ++ show got ++ ", want " ++ show want)

isTrue :: String -> Bool -> IO ()
isTrue what got = unless got $ assertFail (what ++ ": expected True")

isFalse :: String -> Bool -> IO ()
isFalse what got = when got $ assertFail (what ++ ": expected False")

containsElem :: (Eq a, Show a) => String -> [a] -> a -> IO ()
containsElem what haystack needle =
  unless (needle `elem` haystack) $
    assertFail (what ++ ": " ++ show needle ++ " not found in " ++ show haystack)

-- ---------------------------------------------------------------------------
-- main
-- ---------------------------------------------------------------------------

main :: IO ()
main = do
  -- Check 4 prints non-ASCII on failure; without this the handle's encoding
  -- follows the locale and a C-locale CI box dies with "invalid character"
  -- while REPORTING a failure, which is a maddening way to lose a diagnostic.
  hSetEncoding stdout utf8
  putStrLn "=== htmlsanitizer Haskell binding conformance ==="
  v <- abiVersion
  putStrLn ("engine: ABI v" ++ show v)

  fs <- newIORef []
  runChecks fs
  failures <- readIORef fs

  putStrLn ("=== " ++ show (length failures) ++ " failed ===")
  if null failures
    then exitSuccess
    else do
      putStrLn "failures:"
      forM_ failures (\f -> putStrLn ("  - " ++ f))
      exitFailure

runChecks :: Failures -> IO ()
runChecks fs = do
  -- ---- the twelve ----

  check fs "01 script removed" $ \s -> do
    out <- sanitize s "<div>Hello <script>alert(1)</script> world!</div>"
    eqStr "output" out "<div>Hello  world!</div>"

  check fs "02 onclick removed" $ \s -> do
    out <- sanitize s "<div onclick=\"alert(1)\">Hello</div>"
    eqStr "output" out "<div>Hello</div>"

  check fs "03 empty string" $ \s -> do
    -- The classic NULL-vs-"" bug: a binding that maps a null char* to an
    -- error, or that returns a garbage pointer for the empty result.
    out <- sanitize s ""
    eqStr "output" out ""

  check fs "04 utf-8 round trip" $ \s -> do
    -- These are UTF-8 byte literals via OverloadedStrings on ByteString,
    -- which takes the low 8 bits of each Char — so we spell the multibyte
    -- sequences explicitly rather than trusting the literal.
    let cafe = B.concat ["<div>caf", B.pack [0xc3, 0xa9], " ", B.pack [0xe2, 0x98, 0x95], "</div>"]
    out <- sanitize s cafe
    eqStr "output" out cafe

  check fs "05 allow custom tag" $ \s -> do
    before <- sanitize s "<my-widget>x</my-widget>"
    eqStr "before" before ""
    ok <- allow s Tags "my-widget"
    isTrue "allow returned True" ok
    after <- sanitize s "<my-widget>x</my-widget>"
    eqStr "after" after "<my-widget>x</my-widget>"

  check fs "06 disallow tag" $ \s -> do
    before <- sanitize s "<div>x</div>"
    eqStr "before" before "<div>x</div>"
    ok <- disallow s Tags "div"
    isTrue "disallow returned True" ok
    after <- sanitize s "<div>x</div>"
    eqStr "after" after ""

  check fs "07 membership and count" $ \s -> do
    http <- isAllowed s Schemes "http"
    isTrue "http allowed" http
    gopher <- isAllowed s Schemes "gopher"
    isFalse "gopher allowed" gopher
    n <- countList s Schemes
    eqInt "scheme count" n 2

  check fs "08 enumeration" $ \s -> do
    got <- sortedItems s Schemes
    eqInt "enumerated count" (length got) 2
    isTrue "sorted == [http, https]" (got == ["http", "https"])

  check fs "09 keep child nodes" $ \s -> do
    before <- sanitize s "<div><nope>Hello <span>world</span></nope></div>"
    eqStr "before" before "<div></div>"
    setKeepChildNodes s True
    flag <- getKeepChildNodes s
    isTrue "keepChildNodes reads back True" flag
    after <- sanitize s "<div><nope>Hello <span>world</span></nope></div>"
    eqStr "after" after "<div>Hello <span>world</span></div>"

  -- Check 10 and 11 are the ones worth being careful about — the callback
  -- trampoline and the string-returning hook. Haskell expresses both natively
  -- via `foreign import ccall "wrapper"`, so this binding skips neither.

  check fs "10 onRemovingTag cancels" $ \s -> do
    seen <- newIORef ([] :: [(B.ByteString, N.Reason)])
    onRemovingTag s $ Just $ \node reason -> do
      name <- nodeName node
      modifyIORef' seen (++ [(name, reason)])
      pure (name == "keep-me")
    out <- sanitize s "<div><keep-me>a</keep-me><drop-me>b</drop-me></div>"
    eqStr "output" out "<div><keep-me>a</keep-me></div>"
    got <- readIORef seen
    containsElem "hook saw keep-me" (map fst got) "keep-me"
    containsElem "hook saw drop-me" (map fst got) "drop-me"
    -- An int/long width mismatch in the callback signature shows up here as a
    -- garbage reason rather than NotAllowedTag.
    containsElem "reason is a real int" (map snd got) N.NotAllowedTag

  check fs "11 onFilterUrl rewrites" $ \s -> do
    onFilterUrl s $ Just $ \_elem _raw resolved ->
      pure $
        if resolved == "https://example.com/logo.png"
          then "https://cdn.example.net/logo.png"
          else resolved
    out <- sanitizeWithBase s "<img src=\"logo.png\">" "https://example.com"
    eqStr "output" out "<img src=\"https://cdn.example.net/logo.png\">"

  checkIO fs "12 handles are independent" $
    withSanitizer $ \a -> withSanitizer $ \b -> do
      _ <- allow a Tags "only-in-a"
      inA <- isAllowed a Tags "only-in-a"
      inB <- isAllowed b Tags "only-in-a"
      isTrue "a knows the tag" inA
      isFalse "b does not" inB

  -- ---- a few extras that exercise the remaining callback shapes ----

  check fs "onRemovingAttribute sees the attribute" $ \s -> do
    seen <- newIORef ([] :: [B.ByteString])
    onRemovingAttribute s $ Just $ \el attr _reason -> do
      en <- nodeName el
      an <- attrName attr
      av <- attrValue attr
      modifyIORef' seen (++ [B.intercalate "/" [en, an, av]])
      pure False -- False = proceed with the removal
    out <- sanitize s "<div onclick=\"alert(1)\">x</div>"
    eqStr "output" out "<div>x</div>"
    got <- readIORef seen
    containsElem "attribute hook" got "div/onclick/alert(1)"

  check fs "onRemovingComment cancels" $ \s -> do
    onRemovingComment s $ Just $ \_node -> pure True
    out <- sanitize s "<div>a<!-- keep -->b</div>"
    eqStr "output" out "<div>a<!-- keep -->b</div>"

  check fs "onRemovingStyle is four-arg" $ \s -> do
    seen <- newIORef ([] :: [(B.ByteString, B.ByteString)])
    onRemovingStyle s $ Just $ \_el name value _reason -> do
      modifyIORef' seen (++ [(name, value)])
      pure (name == "-custom-thing")
    out <- sanitize s "<div style=\"-custom-thing: 3; color: red\">x</div>"
    isTrue "custom property kept" ("-custom-thing" `B.isInfixOf` out)
    got <- readIORef seen
    containsElem "style hook args" got ("-custom-thing", "3")

  check fs "onPostProcessNode visits" $ \s -> do
    kinds <- newIORef ([] :: [N.NodeKind])
    onPostProcessNode s $ Just $ \node -> do
      k <- nodeKind node
      modifyIORef' kinds (++ [k])
    _ <- sanitize s "<div><span>a</span><span>b</span></div>"
    got <- readIORef kinds
    isTrue "onPostProcessNode fired" (not (null got))

  check fs "node tree navigation" $ \s -> do
    kindRef <- newIORef (N.UnknownKind 0)
    childRef <- newIORef (0 :: Int)
    parentOk <- newIORef False
    onPostProcessDom s $ Just $ \doc -> do
      k <- nodeKind doc
      writeIORef kindRef k
      kids <- nodeChildren doc
      writeIORef childRef (length kids)
      case kids of
        (first : _) -> do
          p <- nodeParent first
          writeIORef parentOk (maybe False (const True) p)
        [] -> pure ()
    _ <- sanitize s "<div>a</div><p>b</p>"
    k <- readIORef kindRef
    isTrue "document kind" (k == N.Document)
    n <- readIORef childRef
    isTrue "document has >= 2 children" (n >= 2)
    po <- readIORef parentOk
    isTrue "child's parent is set" po

  check fs "setAttrValue rewrites an attribute in place" $ \s -> do
    onRemovingAttribute s $ Just $ \_el attr _reason -> do
      an <- attrName attr
      when (an == "onclick") $ do
        setAttrValue attr "sanitised"
        v <- attrValue attr
        eqStr "value after setAttrValue" v "sanitised"
      pure False
    out <- sanitize s "<div onclick=\"alert(1)\">x</div>"
    eqStr "output" out "<div>x</div>"

  check fs "attribute enumeration" $ \s -> do
    names <- newIORef ([] :: [B.ByteString])
    onPostProcessNode s $ Just $ \node -> do
      k <- nodeKind node
      nm <- nodeName node
      when (k == N.Element && nm == "a") $ do
        attrs <- nodeAttributes node
        forM_ attrs $ \a -> do
          an <- attrName a
          modifyIORef' names (++ [an])
    _ <- sanitize s "<a href=\"https://example.com/\" title=\"t\">x</a>"
    got <- readIORef names
    containsElem "href seen" got "href"
    containsElem "title seen" got "title"

  check fs "clearing a hook restores default behaviour" $ \s -> do
    onRemovingTag s $ Just $ \node _r -> (== "keep-me") <$> nodeName node
    kept <- sanitize s "<div><keep-me>a</keep-me></div>"
    eqStr "with hook" kept "<div><keep-me>a</keep-me></div>"
    onRemovingTag s Nothing
    dropped <- sanitize s "<div><keep-me>a</keep-me></div>"
    eqStr "hook cleared" dropped "<div></div>"

  check fs "re-registering a hook does not crash or double-free" $ \s -> do
    -- Each registration mints a new FunPtr stub; the superseded ones stay
    -- retained until close. If the binding freed the old stub eagerly, the
    -- engine's next call would jump through reclaimed memory — which shows up
    -- here as a segfault, not a failed assertion.
    onRemovingTag s $ Just $ \node _r -> (== "keep-me") <$> nodeName node
    onRemovingTag s $ Just $ \node _r -> (== "keep-me") <$> nodeName node
    out <- sanitize s "<div><keep-me>a</keep-me></div>"
    eqStr "second registration wins" out "<div><keep-me>a</keep-me></div>"
    clearHooks s
    after <- sanitize s "<div><keep-me>a</keep-me></div>"
    eqStr "clearHooks cleared it" after "<div></div>"

  check fs "sanitizeDocument is wired" $ \s -> do
    out <- sanitizeDocument s "<div>doc<script>x</script></div>"
    eqStr "output" out "<html><head></head><body><div>doc</div></body></html>"

  check fs "allowDataAttributes flag" $ \s -> do
    off <- getAllowDataAttributes s
    isFalse "off by default" off
    setAllowDataAttributes s True
    on <- getAllowDataAttributes s
    isTrue "on after set" on
    out <- sanitize s "<div data-x=\"1\"></div>"
    eqStr "output" out "<div data-x=\"1\"></div>"

  check fs "clear empties a policy list" $ \s -> do
    ok <- clearList s Schemes
    isTrue "clear returned True" ok
    n <- countList s Schemes
    eqInt "count after clear" n 0

  check fs "item at an out-of-range index is empty" $ \s -> do
    oob <- itemAt s Schemes 999
    eqStr "out of range" oob ""
    neg <- itemAt s Schemes (-1)
    eqStr "negative index" neg ""

  check fs "every policy list is reachable" $ \s -> do
    -- Proves the `which` selector constants line up with the engine's, in all
    -- six positions — an off-by-one here would silently edit the wrong list.
    forM_ [minBound .. maxBound :: Which] $ \w -> do
      _ <- allow s w "probe-item"
      hit <- isAllowed s w "probe-item"
      isTrue ("selector " ++ show w ++ " round-trips") hit

  check fs "items enumerates in the engine's own order" $ \s -> do
    raw <- items s Schemes
    eqInt "raw count" (length raw) 2
    isTrue "sorting it gives [http, https]" (sort raw == ["http", "https"])

  check fs "ABI version" $ \_s -> do
    v <- abiVersion
    isTrue "abiVersion >= 1" (v >= 1)

  checkIO fs "closed sanitizer rejects use" $ do
    s <- new
    close s
    closed <- isClosed s
    isTrue "isClosed after close" closed
    r <- try (sanitize s "<div>x</div>") :: IO (Either SomeException B.ByteString)
    case r of
      Left _ -> pure ()
      Right _ -> assertFail "expected an exception from a closed sanitizer"
    close s -- idempotent

  checkIO fs "many sanitize calls do not leak or crash" $
    withSanitizer $ \s -> do
      -- A returned char* that was never freed would show up here as steadily
      -- growing RSS; a double free would crash. Cheap insurance over the
      -- takeString contract, and it also exercises a registered hook being
      -- entered thousands of times from a `safe` foreign call.
      onPostProcessNode s $ Just $ \_node -> pure ()
      let payload =
            B.concat
              ["<div onclick=\"x\">caf", B.pack [0xc3, 0xa9], " <script>no</script></div>"]
      forM_ [1 :: Int .. 5000] $ \_ -> sanitize s payload
      out <- sanitize s "<div>ok</div>"
      eqStr "still working" out "<div>ok</div>"

File: lua/test/conformance.lua
--- The 12-check binding conformance suite (docs/conformance.md).
---
--- Proves the Lua binding marshals every value shape across the FFI. It is
--- NOT a sanitizer test suite — the behavioural cases live in the sanitizer core's own
--- tests and run once, in Aether.
---
--- Lua 5.4 has no de-facto-standard test framework in its distribution, so
--- this is a plain assertion runner: no dependency to install, and the exit
--- code is the result. Run it with `lua/run_tests.sh`, or:
---
---     lua5.4 -e 'package.cpath="./?.so;"..package.cpath' \
---            -e 'package.path="./src/?.lua;"..package.path' \
---            test/conformance.lua

local hs = require("htmlsanitizer")

-- ---- a minimal test harness ----

local passed, failed = 0, 0
local failures = {}

local function record_failure(name, detail)
  failed = failed + 1
  failures[#failures + 1] = name .. "\n      " .. detail
  print(string.format("  FAIL %s\n      %s", name, detail))
end

local function test(name, body)
  local s = hs.new()
  local ok, err = pcall(body, s)
  s:close()
  if ok then
    passed = passed + 1
    print("  PASS " .. name)
  else
    record_failure(name, tostring(err))
  end
end

local function eq(got, want, what)
  if got ~= want then
    error(string.format("%s:\n        got  %q\n        want %q",
      what or "value", tostring(got), tostring(want)), 2)
  end
end

local function is_true(got, what)
  if not got then error((what or "value") .. ": expected true, got " ..
    tostring(got), 2) end
end

local function is_false(got, what)
  if got then error((what or "value") .. ": expected false, got " ..
    tostring(got), 2) end
end

local function contains(list, want, what)
  for _, v in ipairs(list) do
    if v == want then return end
  end
  error(string.format("%s: %q not found in {%s}", what or "list",
    tostring(want), table.concat(list, ", ")), 2)
end

print("=== htmlsanitizer Lua binding conformance ===")
print(string.format("sanitizer core: %s (ABI v%d)", hs.engine_path(), hs.abi_version()))

-- ---- the twelve ----

test("01 script removed", function(s)
  eq(s:sanitize("<div>Hello <script>alert(1)</script> world!</div>"),
     "<div>Hello  world!</div>")
end)

test("02 onclick removed", function(s)
  eq(s:sanitize('<div onclick="alert(1)">Hello</div>'), "<div>Hello</div>")
end)

test("03 empty string", function(s)
  eq(s:sanitize(""), "")
end)

test("04 utf-8 round trip", function(s)
  -- Lua strings are byte strings; this proves the bytes survive intact rather
  -- than being mangled by a Latin-1 assumption anywhere in the chain.
  eq(s:sanitize("<div>café ☕</div>"), "<div>café ☕</div>")
end)

test("05 allow custom tag", function(s)
  eq(s:sanitize("<my-widget>x</my-widget>"), "")
  s.allowed_tags:add("my-widget")
  eq(s:sanitize("<my-widget>x</my-widget>"), "<my-widget>x</my-widget>")
end)

test("06 disallow tag", function(s)
  eq(s:sanitize("<div>x</div>"), "<div>x</div>")
  s.allowed_tags:remove("div")
  eq(s:sanitize("<div>x</div>"), "")
end)

test("07 membership and count", function(s)
  is_true(s.allowed_schemes:contains("http"), "http allowed")
  is_false(s.allowed_schemes:contains("gopher"), "gopher allowed")
  eq(s.allowed_schemes:count(), 2, "scheme count")
  eq(#s.allowed_schemes, 2, "scheme count via __len")
end)

test("08 enumeration", function(s)
  local got = s.allowed_schemes:sorted()
  eq(#got, 2, "enumerated count")
  eq(got[1], "http")
  eq(got[2], "https")
end)

test("09 keep child nodes", function(s)
  eq(s:sanitize("<div><nope>Hello <span>world</span></nope></div>"),
     "<div></div>")
  s:set_keep_child_nodes(true)
  is_true(s:get_keep_child_nodes(), "keep_child_nodes")
  eq(s:sanitize("<div><nope>Hello <span>world</span></nope></div>"),
     "<div>Hello <span>world</span></div>")
end)

test("10 on_removing_tag cancels", function(s)
  local names, reasons = {}, {}
  s:on_removing_tag(function(node, reason)
    names[#names + 1] = node:name()
    reasons[#reasons + 1] = reason
    return node:name() == "keep-me"
  end)
  eq(s:sanitize("<div><keep-me>a</keep-me><drop-me>b</drop-me></div>"),
     "<div><keep-me>a</keep-me></div>")
  contains(names, "keep-me", "hook saw keep-me")
  contains(names, "drop-me", "hook saw drop-me")
  eq(reasons[1], hs.REASON_NOT_ALLOWED_TAG, "reason is an int, not garbage")
end)

test("11 on_filter_url rewrites", function(s)
  s:on_filter_url(function(_elem, _raw, resolved)
    if resolved == "https://example.com/logo.png" then
      return "https://cdn.example.net/logo.png"
    end
    return resolved
  end)
  eq(s:sanitize('<img src="logo.png">', "https://example.com"),
     '<img src="https://cdn.example.net/logo.png">')
end)

test("12 handles are independent", function()
  local a, b = hs.new(), hs.new()
  a.allowed_tags:add("only-in-a")
  is_true(a.allowed_tags:contains("only-in-a"), "a knows the tag")
  is_false(b.allowed_tags:contains("only-in-a"), "b does not")
  a:close()
  b:close()
end)

-- ---- a few extras that exercise the remaining callback shapes ----

test("on_removing_attribute sees the attribute", function(s)
  local seen = {}
  s:on_removing_attribute(function(elem, attr, _reason)
    seen[#seen + 1] = elem:name() .. "/" .. attr:name() .. "/" .. attr:value()
    return false
  end)
  eq(s:sanitize('<div onclick="alert(1)">x</div>'), "<div>x</div>")
  contains(seen, "div/onclick/alert(1)", "attribute hook")
end)

test("on_removing_comment cancels", function(s)
  s:on_removing_comment(function(_node) return true end)
  eq(s:sanitize("<div>a<!-- keep -->b</div>"), "<div>a<!-- keep -->b</div>")
end)

test("on_removing_style is four-arg", function(s)
  local seen = {}
  s:on_removing_style(function(_elem, name, value, _reason)
    seen[#seen + 1] = name .. "=" .. value
    return name == "-custom-thing"
  end)
  local out = s:sanitize('<div style="-custom-thing: 3; color: red">x</div>')
  is_true(out:find("-custom-thing", 1, true) ~= nil, "custom property kept")
  contains(seen, "-custom-thing=3", "style hook args")
end)

test("post_process_node visits", function(s)
  local kinds = {}
  s:on_post_process_node(function(node) kinds[#kinds + 1] = node:kind() end)
  s:sanitize("<div><span>a</span><span>b</span></div>")
  is_true(#kinds > 0, "post_process_node fired")
end)

test("node tree navigation", function(s)
  local kind, children
  s:on_post_process_dom(function(doc)
    kind = doc:kind()
    children = doc:child_count()
    -- walk one level, proving child_at/parent are wired
    if children > 0 then
      local first = doc:child_at(1)
      assert(first ~= nil, "child_at(1) returned nil")
      assert(first:parent() ~= nil, "parent() returned nil")
    end
  end)
  s:sanitize("<div>a</div><p>b</p>")
  eq(kind, hs.NODE_DOCUMENT, "document kind")
  is_true(children >= 2, "document has >= 2 children")
end)

test("attr_set_value rewrites in place", function(s)
  s:on_removing_attribute(function(_elem, attr, _reason)
    if attr:name() == "onclick" then
      attr:set_value("sanitised")
      eq(attr:value(), "sanitised", "value after set_value")
    end
    return false
  end)
  eq(s:sanitize('<div onclick="alert(1)">x</div>'), "<div>x</div>")
end)

test("attributes helper enumerates", function(s)
  local names = {}
  s:on_post_process_node(function(node)
    if node:kind() == hs.NODE_ELEMENT and node:name() == "a" then
      for _, a in ipairs(hs.attributes(node)) do
        names[#names + 1] = a:name()
      end
    end
  end)
  s:sanitize('<a href="https://example.com/" title="t">x</a>')
  contains(names, "href", "href seen")
  contains(names, "title", "title seen")
end)

test("clearing a hook restores default behaviour", function(s)
  s:on_removing_tag(function(node, _r) return node:name() == "keep-me" end)
  eq(s:sanitize("<div><keep-me>a</keep-me></div>"),
     "<div><keep-me>a</keep-me></div>")
  s:on_removing_tag(nil)
  eq(s:sanitize("<div><keep-me>a</keep-me></div>"), "<div></div>")
end)

test("sanitize_document is wired", function(s)
  eq(s:sanitize_document("<div>doc<script>x</script></div>"), "<html><head></head><body><div>doc</div></body></html>")
end)

test("allow_data_attributes flag", function(s)
  is_false(s:get_allow_data_attributes(), "off by default")
  s:set_allow_data_attributes(true)
  is_true(s:get_allow_data_attributes(), "on after set")
  eq(s:sanitize('<div data-x="1"></div>'), '<div data-x="1"></div>')
end)

test("clear empties a policy list", function(s)
  s.allowed_schemes:clear()
  eq(s.allowed_schemes:count(), 0, "count after clear")
end)

test("item_at out of range is an empty string", function(s)
  eq(s.allowed_schemes:at(999), "", "out of range")
  eq(s.allowed_schemes:at(0), "", "below range (1-based)")
end)

test("abi version", function(s)
  is_true(s:abi_version() >= 1, "abi_version >= 1")
end)

test("closed sanitizer rejects use", function()
  local s = hs.new()
  s:close()
  local ok, err = pcall(function() return s:sanitize("<div>x</div>") end)
  is_false(ok, "sanitize on a closed handle")
  is_true(tostring(err):find("closed", 1, true) ~= nil,
    "error mentions 'closed': " .. tostring(err))
  s:close() -- idempotent
end)

test("a retained node errors instead of reading freed memory", function(s)
  local escaped
  s:on_post_process_dom(function(doc) escaped = doc end)
  s:sanitize("<div>a</div>")
  is_true(escaped ~= nil, "captured a node")
  local ok, err = pcall(function() return escaped:name() end)
  is_false(ok, "using a node after sanitize() returned")
  is_true(tostring(err):find("no longer valid", 1, true) ~= nil,
    "error explains the borrow: " .. tostring(err))
end)

test("one-shot helpers", function()
  eq(hs.sanitize("<div>a<script>b</script></div>"), "<div>a</div>")
  eq(hs.sanitize_document("<div>a<script>b</script></div>"), "<html><head></head><body><div>a</div></body></html>")
end)

-- ---- result ----

print(string.format("=== %d passed, %d failed ===", passed, failed))
if failed > 0 then
  print("failures:")
  for _, f in ipairs(failures) do print("  - " .. f) end
  os.exit(1)
end
os.exit(0)
